web-bundle/src/main/java/com/graphhopper/customspeeds/CustomSpeedsUtils.java

```diff
@@ -76,37 +81,76 @@ private static Map<String, File> getVehicleNameToCustomSpeedFile(List<Profile> p
         try {
             Reader in = new FileReader(customSpeedFile);
             // setHeader() with no args allows the column names to be automatically parsed from the first line of the
             // file
             CSVParser parser = CSVFormat.Builder.create().setHeader().build().parse(in);
+            directionalCustomSpeedsProvided = parser.getHeaderNames().contains(CUSTOM_SPEED_BWD_COL_NAME);
             for (CSVRecord record : parser) {
                 Long osmWayId = Long.parseLong(record.get(CUSTOM_SPEED_OSM_WAY_ID_COL_NAME));
                 Double maxSpeed = Double.parseDouble(record.get(CUSTOM_SPEED_MAX_SPEED_KPH_COL_NAME));
-                osmWayIdToMaxSpeed.put(osmWayId, maxSpeed);
+
+                // If `bwd` column is present, use it to store directional speed for the Way
+                if (directionalCustomSpeedsProvided) {
+                    String bwdString;
+                    try {
+                        bwdString = record.get(CUSTOM_SPEED_BWD_COL_NAME);
+                    } catch (IllegalArgumentException e) {
+                        throw new IllegalArgumentException("bwd column was included in custom speed file, but no value was provided for OSM Way " + osmWayId);
+                    }
+                    Boolean bwd = Boolean.parseBoolean(bwdString);
+                    osmWayIdAndBwdToMaxSpeed.put(Pair.of(osmWayId, bwd), maxSpeed);
+                }
+                // Otherwise, store the speed for the way in both directions
+                else {
+                    osmWayIdAndBwdToMaxSpeed.put(Pair.of(osmWayId, Boolean.TRUE), maxSpeed);
+                    osmWayIdAndBwdToMaxSpeed.put(Pair.of(osmWayId, Boolean.FALSE), maxSpeed);
+                }
             }
             parser.close();
         } catch (IOException e) {
             throw new RuntimeException("Failed to parse custom speed file at path " + customSpeedFile.getAbsolutePath()
                     + ". Please ensure file exists and is in the correct format!", e);
         }
 
-        return osmWayIdToMaxSpeed.build();
+        return Pair.of(osmWayIdAndBwdToMaxSpeed.build(), directionalCustomSpeedsProvided);
     }
 
-    public static Optional<Double> getCustomMaxSpeed(ReaderWay way, ImmutableMap<Long, Double> osmWayIdToCustomMaxSpeed) {
+    public static Optional<Double> getCustomMaxSpeed(ReaderWay way, ImmutableMap<Pair<Long, Boolean>, Double> osmWayIdAndBwdToCustomMaxSpeed, boolean bwd) {
         // n.b. CarAverageSpeedParser sets max speed to be 90% of the OSM's max speed for the way, but we don't apply the 90%
         // discount for the custom speeds we've been explicitly given
-        return Optional.ofNullable(osmWayIdToCustomMaxSpeed.get(way.getId()));
+        return Optional.ofNullable(osmWayIdAndBwdToCustomMaxSpeed.get(Pair.of(way.getId(), bwd)));
     }
 
-    public static Optional<Double> getCustomBadSurfaceSpeed(ReaderWay way, ImmutableMap<Long, Double> osmWayIdToCustomMaxSpeed) {
```

Comment:
I looked through where `getBadSurfaceSpeed()` (which we've overridden with this custom method) is actually called, and it turns out it's not really having an effect right now, so I think we can get rid of it. It's [only called](https://github.com/graphhopper/graphhopper/blob/master/core/src/main/java/com/graphhopper/routing/util/parsers/CarAverageSpeedParser.java#L134) in `CarAccessParser.handleWayTags()`, and right after it's called, `applyMaxSpeed()` is called (which we've customized to apply custom speeds, if they exist()